django/contrib/gis/db/backends/postgis/base.py

```diff
@@ -1,17 +1,106 @@
+from functools import lru_cache
+
 from django.db.backends.base.base import NO_DB_ALIAS
-from django.db.backends.postgresql.base import (
-    DatabaseWrapper as Psycopg2DatabaseWrapper,
-)
+from django.db.backends.postgresql.base import DatabaseWrapper as PsycopgDatabaseWrapper
+from django.db.backends.postgresql.psycopg_any import is_psycopg3
 
+from .adapter import PostGISAdapter
 from .features import DatabaseFeatures
 from .introspection import PostGISIntrospection
 from .operations import PostGISOperations
 from .schema import PostGISSchemaEditor
 
+if is_psycopg3:
+    from psycopg.adapt import Dumper, Loader
+    from psycopg.pq import Format
+    from psycopg.types import TypeInfo
+
+    class GeometryType:
+        pass
+
+    class GeographyType:
+        pass
+
+    class RasterType:
+        pass
+
+    class BinaryLoader(Loader):
+        format = Format.BINARY
+
+        def load(self, data):
+            if not isinstance(data, bytes):
+                data = bytes(data)
+            return data
+
+    class TextLoader(Loader):
+        def load(self, data):
+            if isinstance(data, memoryview):
+                return bytes(data).decode()
+            return data.decode()
```

Comment:
Can we use `psycopg.types.string.TextLoader` instead?

Reply:
Not 100% sure, could you try it? Looking at the two classes in comparison I fear they might not be exactly the same.

Reply:
Especially for `BinaryLoader` where we return bytes, `TextBinaryLoader` will afaiks decode due to the fact that we set the conn encoding to utf8?

Reply:
I see you adjusted it already, I have started a full testrun here: https://github.com/apollo13/django/actions/runs/3686577542

Reply:
Interesting, I do not understand why the builtin loaders work, but I haven't reevaluated that code that often after writing it initially. So jay for less code :+1: